cloud/terraform/templates.go

```diff
@@ -321,3 +340,452 @@ done
 
 {{- end }} {{/* end configure */}}
 `
+
+const podNetworkCNICalico = `
+# This manifest is forked from https://docs.projectcalico.org/v3.0/getting-started/kubernetes/installation/hosted/kubeadm/1.7/calico.yaml
+# The ipam is changed from "calico-ipam" to "host-local".
+
+# Calico Version v3.0.7
+# https://docs.projectcalico.org/v3.0/releases#v3.0.7
+# This manifest includes the following component versions:
+#   calico/node:v3.0.7
+#   calico/cni:v2.0.5
+#   calico/kube-controllers:v2.0.4
+
+# This ConfigMap is used to configure a self-hosted Calico installation.
+kind: ConfigMap
+apiVersion: v1
+metadata:
+  name: calico-config
+  namespace: kube-system
+data:
+  # The location of your etcd cluster.  This uses the Service clusterIP
+  # defined below.
+  etcd_endpoints: "http://10.96.232.136:6666"
+
+  # Configure the Calico backend to use.
+  calico_backend: "bird"
+
+  # The CNI network configuration to install on each node.
+  cni_network_config: |-
+    {
+      "name": "k8s-pod-network",
+      "cniVersion": "0.3.0",
+      "plugins": [
+        {
+          "type": "calico",
+          "etcd_endpoints": "__ETCD_ENDPOINTS__",
+          "log_level": "info",
+          "mtu": 1500,
+          "ipam": {
+              "type": "host-local",
+              "subnet": "usePodCidr"
```

Comment:
@karan we change it from "calico-ipam" to "host-local".
Yes, actually I prefer to install from web. But I didn't find the host-local yaml from calico website.
That's why I checkedin the yaml in the previous commit. So my question is, do you have any recommendation for store the yaml instead of store it in the github repo? Create an google cloud storage project for it? What stop us from storing the yaml in git?

Reply:
A consideration here is that this requires additional internet access to deploy the cluster - at least this should be documented.

A thought is that the URL is configurable - so we could download calico from anywhere, being some place we might store it (for example, we might put it alongside other binaries that the on-prem installation can access) or even a private local URL.

Reply:
@satyasm and I also have the thought about the internet access, but in that case, we need to pull image online unless we provide private image repo in the on-perm environment. 
I agree that if we can ship the yaml/config files with binaries is a good choice, but the current design is to use the template file to ship the setup scripts, I would assume that the design prefers to ship only one binary.

Reply:
How about add a CNI section to ClusterNetwork (https://github.com/cadmuxe/cluster-api/blob/master/pkg/apis/cluster/v1alpha1/cluster_types.go#L61)
Something like:
cniProvider {
type: "calico/weavenet",
"ProviderConfig": {}   // leave this part to maintainer
}
or we can just use:
cniProvider {
type: "calico/weavenet",
providerSpec: "url"  
}

Reply:
You can do that, but that brings me back to my point of deciding if we want to support multiple CNI or simply swap weave to Calico. If the former, and you want to add a field to cluster spec, that will require an API change -- this will need to be discussed more broadly with the community.

Reply:
yes. right now the goal is to swap the default to calico. I agree that we need a broader discussion o CNI selection and chaining.

Reply:
Yeah, agree. In the PR description I had mentioned that we may want to leave the api change open until we have better context. 
For this PR, cloud we change it back to install from URL and check in the yaml to github repo? Later we can either update the api and, if necessary, create online repo for yamls or ship it with binaries.